pkg/plugins/runtime/k8s/controllers/configmap_controller.go

```diff
@@ -0,0 +1,137 @@
+
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/dns"
+	mesh_k8s "github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/metadata"
+)
+
+// ConfigMapReconciler reconciles a ConfigMap object
+type ConfigMapReconciler struct {
+	kube_client.Client
+	kube_record.EventRecorder
+	Scheme          *kube_runtime.Scheme
+	Log             logr.Logger
+	ResourceManager manager.ResourceManager
+	IPAM            dns.IPAM
+	Persistence     *dns.MeshedPersistence
+	Resolver        dns.DNSResolver
+}
+
+func (r *ConfigMapReconciler) Reconcile(req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	mesh, ok := dns.MeshedConfigKey(req.Name)
+	if !ok {
+		return kube_ctrl.Result{}, nil
+	}
+
+	if err := dns.CreateOrUpdateVIPConfig(r.Persistence, r.ResourceManager, r.Resolver, r.IPAM, mesh); err != nil {
+		return kube_ctrl.Result{}, err
+	}
+
+	return kube_ctrl.Result{}, nil
+}
+
+func (r *ConfigMapReconciler) SetupWithManager(mgr kube_ctrl.Manager) error {
+	for _, addToScheme := range []func(*kube_runtime.Scheme) error{kube_core.AddToScheme, mesh_k8s.AddToScheme} {
+		if err := addToScheme(mgr.GetScheme()); err != nil {
+			return err
+		}
+	}
+	return kube_ctrl.NewControllerManagedBy(mgr).
+		For(&kube_core.ConfigMap{}).
+		Watches(&kube_source.Kind{Type: &kube_core.Service{}}, &kube_handler.EnqueueRequestsFromMapFunc{
+			ToRequests: &ServiceToConfigMapsMapper{Client: mgr.GetClient(), Log: r.Log.WithName("service-to-configmap-mapper")},
+		}).
+		Watches(&kube_source.Kind{Type: &mesh_k8s.Dataplane{}}, &kube_handler.EnqueueRequestsFromMapFunc{
+			ToRequests: &DataplaneToMeshMapper{Client: mgr.GetClient(), Log: r.Log.WithName("dataplane-to-configmap-mapper")},
+		}).
+		Watches(&kube_source.Kind{Type: &mesh_k8s.ExternalService{}}, &kube_handler.EnqueueRequestsFromMapFunc{
+			ToRequests: &ExternalServiceToConfigMapsMapper{Client: mgr.GetClient(), Log: r.Log.WithName("external-service-to-configmap-mapperr")},
+		}).
+		Complete(r)
+}
+
+type ServiceToConfigMapsMapper struct {
+	kube_client.Client
+	Log logr.Logger
+}
+
+func (m *ServiceToConfigMapsMapper) Map(obj kube_handler.MapObject) []kube_reconile.Request {
+	cause, ok := obj.Object.(*kube_core.Service)
+	if !ok {
+		m.Log.WithValues("dataplane", obj.Meta).Error(errors.Errorf("wrong argument type: expected %T, got %T", cause, obj.Object), "wrong argument type")
+		return nil
+	}
+
+	ctx := context.Background()
+	svcName := fmt.Sprintf("%s/%s", cause.Namespace, cause.Name)
+	// List Pods in the same namespace
+	pods := &kube_core.PodList{}
+	if err := m.Client.List(ctx, pods, kube_client.InNamespace(obj.Meta.GetNamespace())); err != nil {
+		m.Log.WithValues("service", svcName).Error(err, "failed to fetch Dataplanes in namespace")
+		return nil
+	}
+
+	meshSet := map[string]bool{}
+	for _, pod := range pods.Items {
+		meshSet[pod.Annotations[metadata.KumaMeshAnnotation]] = true
```

Comment:
what if `pod.Annotations[metadata.KumaMeshAnnotation]` is `""` can happen when Pod is excluded from the injection

Reply:
Yes, you are right, thanks! Fixed that